### DISCUSSION TITLE
gRPC (or gRPC-like) protocol for mobile to backend communications

### DISCUSSION THREAD

**Discussion by gitrema** [MobileNativeFoundation/discussions#213] [General]
Hello there,

I am Gianni, an Engineer working at Spotify focusing on the client networking stack and more in general on client to backend communication.

Our mobile app features communicate with the backend by making HTTP requests and JSON as data serialization format primarily.
Our backend is completely moving towards gRPC protocol for micro services communication within the Company network. 
We have a transcoding service (as an envoy sidecar) to transcode HTTP requests coming from clients into gRPC requests and vice-versa for the exposed micro services that have a gRPC API.
Given the direction backend micro services are moving towards we are evaluating the RPC semantic end to end from client to backend in order to reuse the protobuf schema as much as possible.
We have evaluated the use of native gRPC on clients in the past by using [grpc-swift](https://github.com/grpc/grpc-swift) on iOS or [google.io - Kotlin](https://grpc.io/docs/languages/kotlin/) on Android but decided against it for various reasons, mainly because of their binary size and the impact it would have on our mobile apps.

We are interested in using the protobuf schema of the exposed RPC API to generate client code and making the clients send HTTP requests with a protobuf payload that represents a 1:1 mapping to a gRPC request. Has anyone already evaluated libraries such as [ConnectRPC](https://connectrpc.com/) for example?

In our opinion the advantage of using a HTTP approach is the ability to plugin whatever HTTP stack to allow HTTP/3 to be used for example.

What is your preferred choice when it comes to making calls to your backend?
Any interesting challenges or wins you would like to share?

**Comment by 4itdg3x9** [comment]
ConnectRPC is honestly the current gold standard if you want that "gRPC-like" experience without the heavy native baggage. Since it works over standard HTTP/1.1 or HTTP/2, you keep total control over your networking stack—which, as you noted, is critical for features like HTTP/3 and custom connection pooling. The biggest win with the HTTP+Protobuf approach is definitely the shared schema definition. It forces API consistency across the board and lets you use the same [Rainbet Casino Online](https://rainbet.co.com/) protobuf generated code for both mobile and backend. Most teams I’ve seen move this way end up much happier with the performance trade-offs, especially on Android where method count and binary size are always a concern. It’s a very clean bridge between the legacy HTTP world and the new gRPC backend reality.
